extractor/filesystem/language/rust/cargotoml/cargotoml.go

```diff
@@ -0,0 +1,165 @@
+	"github.com/google/osv-scalibr/plugin"
+	"github.com/google/osv-scalibr/purl"
+)
+
+var shaPattern = regexp.MustCompile("^[0-9a-f]{40}$")
+
+type cargoTomlDependency struct {
+	Version string
+	Git     string
+	Rev     string
+}
+
+// UnmarshalTOML parses a dependency from a Cargo.toml file.
+//
+// Dependencies in Cargo.toml can be defined as simple strings (e.g., version)
+// or as more complex objects (e.g., with version, path, etc.)
+//
+// in case both the Version and Git/Path are specified the version should be considered
+// the source of truth
+func (v *cargoTomlDependency) UnmarshalTOML(data any) error {
+	getString := func(m map[string]any, key string) string {
+		v, ok := m[key]
+		if !ok {
+			return ""
+		}
+		s, _ := v.(string)
+		return s
+	}
+
+	switch data := data.(type) {
+	case string:
+		// if the type is string then the data is version
+		v.Version = data
+		return nil
+	case map[string]any:
+		*v = cargoTomlDependency{
+			Version: getString(data, "version"),
+			Git:     getString(data, "git"),
+			Rev:     getString(data, "rev"),
+		}
+		return nil
+	default:
+		return errors.New("invalid format for Cargo.toml dependency")
+	}
+}
+
+// IsCommitSpecified checks if the dependency specifies a Git commit.
+func (v *cargoTomlDependency) IsCommitSpecified() bool {
+	return v.Git != "" && shaPattern.MatchString(v.Rev)
+}
+
+type cargoTomlPackage struct {
+	Name    string `toml:"name"`
+	Version string `toml:"version"`
+}
+
+type cargoTomlFile struct {
+	Package      cargoTomlPackage               `toml:"package"`
+	Dependencies map[string]cargoTomlDependency `toml:"dependencies"`
+}
+
+// Extractor extracts crates.io packages from Cargo.toml files.
+type Extractor struct{}
+
+// Name of the extractor
+func (e Extractor) Name() string { return "rust/Cargotoml" }
+
+// Version of the extractor
+func (e Extractor) Version() int { return 0 }
+
+// FileRequired returns true if the specified file matches Cargo toml file patterns.
+func (e Extractor) FileRequired(api filesystem.FileAPI) bool {
+	return filepath.Base(api.Path()) == "Cargo.toml"
+}
+
+// Requirements of the extractor
+func (e Extractor) Requirements() *plugin.Capabilities {
+	return &plugin.Capabilities{}
+}
+
+// Extract extracts packages from Cargo.toml files passed through the scan input.
+func (e Extractor) Extract(_ context.Context, input *filesystem.ScanInput) ([]*extractor.Inventory, error) {
+	var parsedTomlFile cargoTomlFile
+
+	_, err := toml.NewDecoder(input.Reader).Decode(&parsedTomlFile)
+	if err != nil {
+		return nil, fmt.Errorf("could not extract from %s: %w", input.Path, err)
+	}
+
+	packages := make([]*extractor.Inventory, 0, len(parsedTomlFile.Dependencies)+1)
+
+	packages = append(packages, &extractor.Inventory{
+		Name:      parsedTomlFile.Package.Name,
+		Version:   parsedTomlFile.Package.Version,
+		Locations: []string{input.Path},
+	})
+
+	for name, dependency := range parsedTomlFile.Dependencies {
+		var srcCode *extractor.SourceCodeIdentifier
```

Comment:
Add context cancellation in here like in https://github.com/google/osv-scalibr/blob/99ab001bd6ab7f3fd986d9a33f5aa1da30e18f92/extractor/filesystem/os/dpkg/dpkg.go#L184